src/workerd/io/tracer.c++

```diff
@@ -245,70 +237,84 @@ void WorkerTracer::addSpan(CompleteSpan&& span) {
   }
 
-  trace->bytesUsed = newSize;
-  trace->spans.add(kj::mv(span));
+  // Note: spans will not be shipped to the production version of the legacy tail worker, so we
+  // don't need an exceededSpanLimit variable for it.
+  if (trace->bytesUsed + messageSize > MAX_TRACE_BYTES) {
+    trace->truncated = true;
+  } else {
+    trace->bytesUsed += messageSize;
+    trace->spans.add(kj::mv(span));
+  }
 }
 
 void WorkerTracer::addException(const tracing::InvocationSpanContext& context,
     kj::Date timestamp,
     kj::String name,
     kj::String message,
     kj::Maybe<kj::String> stack) {
-  if (trace->exceededExceptionLimit) {
-    return;
-  }
   // TODO(someday): For now, we're using logLevel == none as a hint to avoid doing anything
   //   expensive while tracing.  We may eventually want separate configuration for exceptions vs.
   //   logs.
   if (pipelineLogLevel == PipelineLogLevel::NONE) {
     return;
   }
-  size_t newSize = trace->bytesUsed + sizeof(tracing::Exception) + name.size() + message.size();
+
+  size_t messageSize = sizeof(tracing::Exception) + name.size() + message.size();
   KJ_IF_SOME(s, stack) {
-    newSize += s.size();
+    messageSize += s.size();
   }
-  if (newSize > MAX_TRACE_BYTES) {
+  KJ_IF_SOME(writer, maybeTailStreamWriter) {
+    if (messageSize <= MAX_TRACE_BYTES) {
+      writer->report(context,
+          {tracing::Exception(timestamp, kj::str(name), kj::str(message), mapCopyString(stack))});
+    }
+  }
+
+  if (trace->exceededExceptionLimit) {
+    return;
+  }
+
+  if (trace->bytesUsed + messageSize > MAX_TRACE_BYTES) {
     trace->exceededExceptionLimit = true;
     trace->truncated = true;
     trace->exceptions.add(timestamp, kj::str("Error"),
         kj::str("Trace resource limit exceeded; subsequent exceptions not recorded."), kj::none);
-    return;
-  }
-  trace->bytesUsed = newSize;
-  KJ_IF_SOME(writer, maybeTailStreamWriter) {
-    writer->report(context,
-        {tracing::Exception(timestamp, kj::str(name), kj::str(message), mapCopyString(stack))});
+  } else {
+    trace->bytesUsed += messageSize;
+    trace->exceptions.add(timestamp, kj::mv(name), kj::mv(message), kj::mv(stack));
   }
-  trace->exceptions.add(timestamp, kj::mv(name), kj::mv(message), kj::mv(stack));
 }
 
 void WorkerTracer::addDiagnosticChannelEvent(const tracing::InvocationSpanContext& context,
     kj::Date timestamp,
     kj::String channel,
     kj::Array<kj::byte> message) {
-  if (trace->exceededDiagnosticChannelEventLimit) {
+  if (pipelineLogLevel == PipelineLogLevel::NONE) {
     return;
   }
-  if (pipelineLogLevel == PipelineLogLevel::NONE) {
+
+  size_t messageSize = sizeof(tracing::DiagnosticChannelEvent) + channel.size() + message.size();
+  KJ_IF_SOME(writer, maybeTailStreamWriter) {
+    if (messageSize <= MAX_TRACE_BYTES) {
+      writer->report(context,
+          {tracing::DiagnosticChannelEvent(
+              timestamp, kj::str(channel), kj::heapArray<kj::byte>(message))});
+    }
```

Comment:
Unlike log and error, which I think would be better to truncate, these I'd say are ok to drop on the floor.

Reply:
Added truncation here too, since it's easy to do the same way as with log/exception.